### PR TITLE
Journal blobpk

### PR DESCRIPTION
**Pull request by ifesdjeen**

### CODE REVIEW

File: src/java/org/apache/cassandra/db/marshal/ByteBufferAccessor.java [hunk 1]
```diff
@@ -318,13 +318,13 @@ public int putFloat(ByteBuffer dst, int offset, float value)
     @Override
     public int putLeastSignificantBytes(ByteBuffer dst, int offset, long register, int bytes)
     {
-        if (dst.remaining() < Long.BYTES)
+        int pos = dst.position() + offset;
+        if (dst.limit() - pos < Long.BYTES)
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/db/marshal/ByteBufferAccessor.java:322]: Should we also add an assert that the number of bytes we want to write is going to fit?

File: src/java/org/apache/cassandra/db/marshal/ValueAccessor.java [hunk 2]
```diff
@@ -525,23 +525,23 @@ public static <V> int putLeastSignificantBytes(ValueAccessor<V> accessor, V dst,
                 break;
             case 3:
                 accessor.putShort(dst, offset, (short)(register >>> 8));
-                accessor.putByte(dst, offset, (byte)register);
+                accessor.putByte(dst, offset + 2, (byte)register);
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/db/marshal/ValueAccessor.java:528]: I was looking at this code several times now, changing subtly because of more edge conditions, maybe it's time to cover this with tests?

File: src/java/org/apache/cassandra/service/accord/AccordKeyspace.java [hunk 3]
```diff
@@ -546,60 +541,38 @@ private static ByteBuffer cellValue(Row row, ColumnMetadata column)
...
 
         public static JournalKey getJournalKey(DecoratedKey key)
         {
-            ByteBuffer[] parts = splitPartitionKey(key);
-            return new JournalKey(getTxnId(parts), getType(parts), getStoreId(parts));
+            ByteBuffer bb = key.getKey();
+            int storeId = ByteBufferAccessor.instance.getUnsignedVInt32(bb, 0);
+            int offset = VIntCoding.readLengthOfVInt(bb, 0);
+            JournalKey.Type type = JournalKey.Type.fromId(bb.get(offset));
+            TxnId txnId = type != JournalKey.Type.COMMAND_DIFF ? TxnId.NONE : CommandSerializers.txnId.deserializeComparable(bb, ByteBufferAccessor.instance, offset + 1);
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/service/accord/AccordKeyspace.java:574]: Should we apply same optimization (i.e. "skipping" txnid bytes and avoiding generating txnid objects) for JournalKeySupport?

File: src/java/org/apache/cassandra/service/accord/AccordKeyspace.java [hunk 4]
```diff
@@ -546,60 +541,38 @@ private static ByteBuffer cellValue(Row row, ColumnMetadata column)
...
+            return VIntCoding.readUnsignedVInt32(pk.getKey(), 0);
         }
 
         public static JournalKey getJournalKey(DecoratedKey key)
         {
-            ByteBuffer[] parts = splitPartitionKey(key);
-            return new JournalKey(getTxnId(parts), getType(parts), getStoreId(parts));
+            ByteBuffer bb = key.getKey();
+            int storeId = ByteBufferAccessor.instance.getUnsignedVInt32(bb, 0);
+            int offset = VIntCoding.readLengthOfVInt(bb, 0);
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/service/accord/AccordKeyspace.java:572]: Should we use `VIntCoding.computeUnsignedVIntSize(storeId);` instead?

File: src/java/org/apache/cassandra/service/accord/JournalKey.java [hunk 5]
```diff
@@ -161,6 +161,11 @@ private void serializeTxnId(TxnId txnId, byte[] out)
             ByteArrayUtil.putInt(out, NODE_OFFSET, txnId.node.id);
         }
 
+        private int serializedSizeOffixedWidthTxnId()
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/service/accord/JournalKey.java]: nit: `Fixed` capitalized?

File: src/java/org/apache/cassandra/service/accord/serializers/CommandSerializers.java [hunk 6]
```diff
@@ -315,216 +316,461 @@ private static int flags(Timestamp executeAt, boolean nullable)
...
-            position += accessor.putLong(dst, position, ts.lsb);
-            position += TopologySerializers.nodeId.serialize(ts.node, dst, accessor, position);
-            int size = position - offset;
-            Preconditions.checkState(size == serializedSize());
-            return size;
+            position += accessor.putByte(dst, position, (byte)encodingFlags);
+            position += accessor.putLeastSignificantBytes(dst, position, epoch, epochLength);
+            position += accessor.putLeastSignificantBytes(dst, position, hlc, hlcLength);
+            position += accessor.putLeastSignificantBytes(dst, position, flags, flagsLength);
+            position += accessor.putLeastSignificantBytes(dst, position, flags, nodeLength);
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/service/accord/serializers/CommandSerializers.java]: `flags` should be node id here?

File: src/java/org/apache/cassandra/service/accord/serializers/CommandSerializers.java [hunk 7]
```diff
@@ -315,216 +316,461 @@ private static int flags(Timestamp executeAt, boolean nullable)
     }
 
     // TODO (expected): optimise using subset serializers, or perhaps simply with some deduping key serializer
-    public static class StoreParticipantsSerializer implements IVersionedSerializer<StoreParticipants>
+    public static class StoreParticipantsSerializer implements UnversionedSerializer<StoreParticipants>
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/service/accord/serializers/CommandSerializers.java:318]: Is the TODO above now considered done?

File: src/java/org/apache/cassandra/service/accord/serializers/CommandSerializers.java [hunk 8]
```diff
@@ -315,216 +316,461 @@ private static int flags(Timestamp executeAt, boolean nullable)
...
-            return StoreParticipants.create(route, owns, executes, waitsOn, touches, hasTouched);
+            Participants<?> superset = KeySerializers.participants.deserialize(in);
+            boolean decodeSubset = superset.domain() == Routable.Domain.Key;
+            if (decodeSubset)
+            {
+                Route<?> route = 0 == (flags & HAS_ROUTE) ? null : 0 != (flags & ROUTE_EQUALS_SUPERSET) ? (Route<?>)superset : KeySerializers.route.deserializeSubset(superset, in);
+                Participants<?> hasTouched = 0 != (flags & HAS_TOUCHED_EQUALS_SUPERSET) ? superset : KeySerializers.participants.deserializeSubset(superset, in);
+                Participants<?> touches = 0 != (flags & TOUCHES_EQUALS_HAS_TOUCHED) ? hasTouched : KeySerializers.participants.deserializeSubset(superset, in);
+                Participants<?> owns = 0 != (flags & OWNS_EQUALS_TOUCHES) ? touches : KeySerializers.participants.deserializeSubset(superset, in);
+                Participants<?> executes = 0 != (flags & EXECUTES_IS_NULL) ? null : 0 != (flags & EXECUTES_IS_OWNS) ? owns : KeySerializers.participants.deserializeSubset(superset, in);
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/service/accord/serializers/CommandSerializers.java:416]: Should we add some flags consistency invariants and check that `EXECUTES_IS_NULL`  is consistent with `EXECUTES_IS_OWNS` and `WAITSON_IS_OWNS` (i.e. they can be true iif `executesIsNull` is false).

File: test/unit/org/apache/cassandra/service/accord/serializers/CommandSerializersTest.java [hunk 9]
```diff
@@ -71,4 +79,48 @@ public void txnIdSerde()
         DataOutputBuffer output = new DataOutputBuffer();
         qt().forAll(AccordGens.txnIds()).check(txnId -> Serializers.testSerde(output, CommandSerializers.txnId, txnId));
     }
+
+    @Test
+    public void txnIdComparable()
+    {
+        qt().withSeed(3447825112572683888L).forAll(AccordGens.txnIds(), AccordGens.txnIds()).check(CommandSerializersTest::testComparable);
```
Comment (ifesdjeen) [inline test/unit/org/apache/cassandra/service/accord/serializers/CommandSerializersTest.java]: Probably worth un-hardcoding the seed.

File: test/distributed/org/apache/cassandra/service/accord/AccordJournalBurnTest.java [hunk 10]
```diff
@@ -124,7 +124,8 @@ public void beforeTest() throws Throwable
     @Test
     public void testOne()
     {
-        long seed = System.nanoTime();
+//        long seed = System.nanoTime();
+        long seed = 1L;
```
Comment (ifesdjeen) [inline test/distributed/org/apache/cassandra/service/accord/AccordJournalBurnTest.java]: probably worth un-hardcoding this too

File: src/java/org/apache/cassandra/utils/vint/VIntCoding.java [hunk 11]
```diff
@@ -308,6 +308,17 @@ public static <V> long getUnsignedVInt(V input, ValueAccessor<V> accessor, int r
         return retval;
     }
 
+    public static <V> int readLengthOfVInt(V input, ValueAccessor<V> accessor, int position)
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/utils/vint/VIntCoding.java:311]: A super fundamental method; should we fuzz it with _a lot_ of different values? Like really go through all byte sizes rather than generating a bunch of longs.

File: test/unit/org/apache/cassandra/service/accord/AccordKeyspaceTest.java [hunk 12]
```diff
@@ -18,6 +18,7 @@
 
 package org.apache.cassandra.service.accord;
 
+import java.io.IOException;
```
Comment (ifesdjeen) [inline test/unit/org/apache/cassandra/service/accord/AccordKeyspaceTest.java]: This is not the only place where we got import problems, so I'll just leave it to you to run checkstyle before commit, since there are a few places.

File: src/java/org/apache/cassandra/service/accord/serializers/IVersionedWithKeysSerializer.java [hunk 13]
```diff
@@ -323,44 +277,132 @@ private void serializeLargeSubset(AbstractRanges serialize, int serializeCount,
             }
         }
 
+        public Routables<?> deserializeSubsetInternal(Routables<?> superset, DataInputPlus in) throws IOException
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/service/accord/serializers/IVersionedWithKeysSerializer.java:231]: I am a bit confused about why this is unused now

File: src/java/org/apache/cassandra/service/accord/serializers/IVersionedWithKeysSerializer.java [hunk 14]
```diff
@@ -323,44 +277,132 @@ private void serializeLargeSubset(AbstractRanges serialize, int serializeCount,
...
+                return result.apply(null, superset);
+            else if (supersetCount >= 64)
+                return result.apply(deserializeLargeRangeSubset(in, superset, supersetCount, (int) encoded), superset);
+            else
+                return result.apply(deserializeSmallRangeSubsetArray(encoded, superset, supersetCount), superset);
+        }
+
         @DontInline
-        private Unseekables<?> deserializeLargeSubset(DataInputPlus in, Unseekables<?> superset, int supersetCount, int delta) throws IOException
+        private Routables<?> deserializeLargeSubset(DataInputPlus in, Routables<?> superset, int supersetCount, int delta) throws IOException
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/service/accord/serializers/IVersionedWithKeysSerializer.java]: This one is also unused; maybe we should at least cover it with tests to make sure it works if/when we decide to use it?

File: src/java/org/apache/cassandra/service/accord/serializers/IVersionedWithKeysSerializer.java [hunk 15]
```diff
@@ -323,44 +277,132 @@ private void serializeLargeSubset(AbstractRanges serialize, int serializeCount,
...
         {
-            int deserializeCount = supersetCount - delta;
             switch (superset.domain())
             {
-                default: throw new AssertionError("Unhandled domain: " + superset.domain());
+                default:
+                    throw new AssertionError("Unhandled domain: " + superset.domain());
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/service/accord/serializers/IVersionedWithKeysSerializer.java]: nit: unhandled enum?

File: src/java/org/apache/cassandra/service/accord/serializers/IVersionedWithKeysSerializer.java
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/service/accord/serializers/IVersionedWithKeysSerializer.java:1]: In this class, now `NullableWithKeysSerializer` is unused.

File: src/java/org/apache/cassandra/service/accord/serializers/IVersionedWithKeysSerializer.java [hunk 16]
```diff
@@ -323,44 +277,132 @@ private void serializeLargeSubset(AbstractRanges serialize, int serializeCount,
...
+        public <T, S extends AbstractUnseekableKeys> T deserializeRoutingKeySubset(S superset, DataInputPlus in, BiFunction<RoutingKey[], S, T> result) throws IOException
+        {
+            long encoded = in.readUnsignedVInt();
+            int supersetCount = superset.size();
+            if (encoded == 0L)
+                return result.apply(null, superset);
+            else if (supersetCount >= 64)
+                return result.apply(deserializeLargeRoutingKeySubset(in, superset, supersetCount, (int) encoded), superset);
+            else
+                return result.apply(deserializeSmallRoutingKeySubsetArray(encoded, superset, supersetCount), superset);
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/service/accord/serializers/IVersionedWithKeysSerializer.java]: should we remove "array" from the method name here?

File: src/java/org/apache/cassandra/service/accord/serializers/KeySerializers.java [hunk 17]
```diff
@@ -387,18 +365,92 @@ public RS deserialize(DataInputPlus in) throws IOException
...
         }
 
+        public RS deserializeSubset(Unseekables<?> superset, DataInputPlus in) throws IOException
+        {
+            byte b = in.readByte();
+            UnseekablesKind kind;
+            RS result;
+            switch (b)
+            {
+                default: throw new IOException("Corrupted input: expected byte 1, 2, 3, 4 or 5; received " + b);
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/service/accord/serializers/KeySerializers.java:375]: "or 6"?

File: src/java/org/apache/cassandra/service/accord/serializers/KeySerializers.java [hunk 18]
```diff
@@ -812,29 +906,115 @@ final void serializeOffsets(KS keys, int startIndex, int endIndex, DataOutputPlu
...
+        }
+
+        @Override
+        public long serializedSize(KS route)
+        {
+            return super.serializedSize(route)
+                   + completeSerializedSize(route);
+        }
+
+        public long serializedSubsetSize(KS route, Routables<?> superset)
```
Comment (ifesdjeen) [inline src/java/org/apache/cassandra/service/accord/serializers/KeySerializers.java:981]: nit: override here and skipSubset

File: src/java/org/apache/cassandra/db/marshal/ByteBufferAccessor.java (hunk 1 above)
Comment (belliottsmith) [reply src/java/org/apache/cassandra/db/marshal/ByteBufferAccessor.java:322]: I'd prefer to leave that to the `BufferOverflowException` to avoid wasting time on very hot path

File: src/java/org/apache/cassandra/db/marshal/ValueAccessor.java (hunk 2 above)
Comment (belliottsmith) [reply src/java/org/apache/cassandra/db/marshal/ValueAccessor.java:528]: I think this version of the method just wasn't used before, as it was totally broken. I think the tests included with this patch should cover its functionality.

File: src/java/org/apache/cassandra/service/accord/AccordKeyspace.java (hunk 3 above)
Comment (belliottsmith) [reply src/java/org/apache/cassandra/service/accord/AccordKeyspace.java:574]: `JournalKeySupport` needs to be fixed width for the moment. When we're able to support variable length keys, for sure.

File: src/java/org/apache/cassandra/service/accord/AccordKeyspace.java (hunk 4 above)
Comment (belliottsmith) [reply src/java/org/apache/cassandra/service/accord/AccordKeyspace.java:572]: This version should be faster, which is why I introduced it. We should probably switch similar call-sites to using this version.

File: src/java/org/apache/cassandra/service/accord/JournalKey.java (hunk 5 above)
Comment (belliottsmith) [reply src/java/org/apache/cassandra/service/accord/JournalKey.java]: doesn't look to be used though...

File: src/java/org/apache/cassandra/utils/vint/VIntCoding.java (hunk 11 above)
Comment (belliottsmith) [reply src/java/org/apache/cassandra/utils/vint/VIntCoding.java:311]: it's copy/paste from the existing VIntCoding methods - which admittedly we should really have direct fuzz testing of, but not sure if this is the best place to start...

File: src/java/org/apache/cassandra/service/accord/serializers/IVersionedWithKeysSerializer.java (hunk 13 above)
Comment (belliottsmith) [reply src/java/org/apache/cassandra/service/accord/serializers/IVersionedWithKeysSerializer.java:231]: this particular caller now uses some of the internal methods this method calls directly. But, I have another patch in flight that may invoke these methods directly, and anyway they may be useful for the future so would prefer not to delete them here.

File: src/java/org/apache/cassandra/service/accord/serializers/IVersionedWithKeysSerializer.java (hunk 14 above)
Comment (belliottsmith) [reply src/java/org/apache/cassandra/service/accord/serializers/IVersionedWithKeysSerializer.java]: I don't think this version needs to survive, will delete

File: src/java/org/apache/cassandra/service/accord/AccordKeyspace.java (hunk 3 above)
Comment (ifesdjeen) [reply src/java/org/apache/cassandra/service/accord/AccordKeyspace.java:574]: Ah, sure: I did not mean to switch it to variable width; only that we might want to avoid de-serializing txn id there too just like here.
